src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1468,15 +1468,21 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
                     memberType = CreateErrorType();
                     member = null;
                     hasErrors = true;
+                    isLengthOrCount = false;
                 }
                 else
                 {
                     member = LookupMembersForPropertyPattern(inputType, expr, diagnostics, ref hasErrors);
                     memberType = member.Type;
+                    isLengthOrCount = memberType.SpecialType == SpecialType.System_Int32 &&
+                                      member.Symbol is { Name: WellKnownMemberNames.LengthPropertyName or WellKnownMemberNames.CountPropertyName, ContainingType: TypeSymbol containingType } memberSymbol &&
+                                      (memberSymbol.Equals(((MethodSymbol?)GetWellKnownTypeMember(WellKnownMember.System_Array__get_Length, BindingDiagnosticBag.Discarded, syntax: node))?.AssociatedSymbol) ||
+                                       TryPerformPatternIndexerLookup(node, containingType, argIsIndex: true, indexerAccess: out _, patternSymbol: out _, out PropertySymbol? lengthProperty, BindingDiagnosticBag.Discarded) &&
+                                       memberSymbol.Equals(lengthProperty));
```

Comment:
:memo: If both Count and Length are present we want the one that makes this type countable.